src/opm/parser/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -331,18 +521,436 @@ namespace Opm {
+
+                    x1 = (X[0]+X[1]+X[4]+X[5])/4.0;
+                    y1 = (Y[0]+Y[1]+Y[4]+Y[5])/4.0;
+
+                    double dy = sqrt(pow((x2-x1), 2.0) + pow((y2-y1), 2.0));
+
+                    // calculate dz
+
+                    z2 = (Z[4]+Z[5]+Z[6]+Z[7])/4.0;
+                    z1 = (Z[0]+Z[1]+Z[2]+Z[3])/4.0;
+
+                    double dz = z2-z1;
+
+                    m_dx_active.push_back(dx);
+                    m_dy_active.push_back(dy);
+                    m_dz_active.push_back(dz);
+
+                    m_depth_active.push_back((z2 + z1) / 2.0);
+                }
+            }
+        }
+    }
+
+    void EclipseGrid::getCellCorners(const std::array<int, 3>& ijk, const std::array<int, 3>& dims,
+                                     std::vector<double>& X,
+                                     std::vector<double>& Y,
+                                     std::vector<double>& Z) const
+    {
+        std::vector<int> zind;
+        std::vector<int> pind;
+
+        if (X.size() < 8 || Y.size() < 8 || Z.size() < 8) {
+            OPM_THROW(std::invalid_argument,
+                      "In routine cellConrner. X, Y and Z should be a vector of size 8");
+        }
+
+        // calculate indices for grid pillars in COORD arrray
+        pind.push_back(ijk[1]*(dims[0]+1)*6 + ijk[0]*6);
+        pind.push_back(pind[0] + 6);
+        pind.push_back(pind[0] + (dims[0]+1)*6);
+        pind.push_back(pind[2] + 6);
+
+        // get depths from zcorn array in ZCORN array
+        zind.push_back(ijk[2]*dims[0]*dims[1]*8 + ijk[1]*dims[0]*4 + ijk[0]*2);
+        zind.push_back(zind[0] + 1);
+        zind.push_back(zind[0] + dims[0]*2);
+        zind.push_back(zind[2] + 1);
+
+        for (int n = 0; n < 4; n++) {
+            zind.push_back(zind[n] + dims[0]*dims[1]*4);
+        }
+
+        for (int n = 0; n< 8; n++) {
+            Z[n] = m_zcorn[zind[n]];
+        }
+
+        for (int  n=0; n<4; n++) {
+            double xt = m_coord[pind[n]];
+            double yt = m_coord[pind[n] + 1];
+            double zt = m_coord[pind[n] + 2];
+
+            double xb = m_coord[pind[n] + 3];
+            double yb = m_coord[pind[n] + 4];
+            double zb = m_coord[pind[n]+5];
+
+            X[n] = xt + (xb-xt) / (zt-zb) * (zt - Z[n]);
+            X[n+4] = xt + (xb-xt) / (zt-zb) * (zt-Z[n+4]);
+
+            Y[n] = yt+(yb-yt)/(zt-zb)*(zt-Z[n]);
+            Y[n+4] = yt+(yb-yt)/(zt-zb)*(zt-Z[n+4]);
+        }
+    }
+
+    std::vector<double> EclipseGrid::makeCoordDxvDyvDzvDepthz(const std::array<int, 3>& dims, const std::vector<double>& DXV, const std::vector<double>& DYV, const std::vector<double>& DZV, const std::vector<double>& DEPTHZ) {
+
+        std::vector<double> coord;
+        coord.reserve((dims[0]+1)*(dims[1]+1)*6);
+
+        for (int j = 0; j < dims[1] + 1; j++) {
+            for (int i = 0; i<dims[0] + 1; i++) {
+
+                double x0 = 0.0;
+
+                for (int n = 0; n < i; n++) {
```

Comment:
This leads to much unnecessary re-calculation (quadratic complexity rather than linear in the I and J dimensions). Better to fill a vector with `std::partial_sum()` outside the loop nest and get the data from there:
```
std::vector<double> x(dims[0] + 1, 0.0);
std::partial_sum(DXV.begin(), DXV.end(), x.begin() + 1);
```
Note the extra element to make it start with zero. Then inside the loop you would do
```
const double x0 = x[i];
```
Same applies to y and z.

Reply:
done